File: go/ontology/dedup.go
// SPDX-License-Identifier: Apache-2.0
package ontology

import (
	"context"
	"fmt"
	"strings"

	"github.com/jeffs-brain/memory/go/llm"
)

// Default thresholds matching the intelligence service configuration.
const (
	// FuzzyLabelThreshold is the default minimum Jaro-Winkler similarity
	// for fuzzy label matching within the same type prefix group.
	FuzzyLabelThreshold = 0.85

	// EmbeddingAutoMerge is the default minimum cosine similarity for
	// automatic semantic merging without human review.
	EmbeddingAutoMerge = 0.9

	// EmbeddingReviewThreshold is the default minimum cosine similarity
	// that flags a type pair for human review.
	EmbeddingReviewThreshold = 0.75
)

// SimilarityFunc is a pluggable string similarity function.
// It accepts two strings and returns a similarity value in [0, 1].
type SimilarityFunc func(a, b string) float64

// DedupMethod describes which deduplication tier matched.
type DedupMethod string

const (
	DedupMethodExact      DedupMethod = "exact"
	DedupMethodFuzzyLabel DedupMethod = "fuzzy_label"
	DedupMethodEmbedding  DedupMethod = "embedding"
)

// DedupResultKind is the discriminant for single-type dedup results.
type DedupResultKind string

const (
	DedupResultExact          DedupResultKind = "exact"
	DedupResultFuzzyMatch     DedupResultKind = "fuzzy_match"
	DedupResultSemanticMatch  DedupResultKind = "semantic_match"
	DedupResultSemanticReview DedupResultKind = "semantic_review"
	DedupResultNew            DedupResultKind = "new"
)

// DedupResult is the outcome of deduplicating a single candidate type
// against an existing registry.
type DedupResult struct {
	Kind         DedupResultKind
	ExistingType *TypeDefinition
	Similarity   float64
}

// MergedPair records a deduplication match between an extracted type
// and an existing type.
type MergedPair struct {
	Extracted     TypeDefinition `json:"extracted"`
	ExistingMatch TypeDefinition `json:"existingMatch"`
	Similarity    float64        `json:"similarity"`
	Method        DedupMethod    `json:"method"`
}

// DeduplicationResult holds the outcome of type deduplication across
// three tiers.
type DeduplicationResult struct {
	Unique           []TypeDefinition `json:"unique"`
	AutoMerged       []MergedPair     `json:"autoMerged"`
	ReviewCandidates []MergedPair     `json:"reviewCandidates"`
}

// DeduplicatorConfig configures the Deduplicator's thresholds and
// similarity algorithm.
type DeduplicatorConfig struct {
	// Embedder provides semantic embeddings. May be nil to skip the
	// semantic tier (exact + fuzzy only).
	Embedder llm.Embedder

	// FuzzyThreshold is the minimum string similarity for fuzzy label
	// matching within the same type prefix group. Default: 0.85.
	FuzzyThreshold float64

	// AutoMergeThreshold is the minimum cosine similarity for automatic
	// semantic merging. Default: 0.9.
	AutoMergeThreshold float64

	// ReviewThreshold is the minimum cosine similarity that flags a pair
	// for human review. Default: 0.75.
	ReviewThreshold float64

	// Similarity is the string similarity function used for fuzzy label
	// matching. Default: JaroWinklerDistance.
	Similarity SimilarityFunc
}

// Deduplicator performs three-tier type deduplication: exact match on
// type ID, fuzzy label match via configurable similarity function
// within the same prefix group, and semantic embedding match via
// cosine similarity.
type Deduplicator struct {
	embedder       llm.Embedder
	fuzzyThreshold float64
	autoMerge      float64
	reviewThreshold float64
	similarity     SimilarityFunc
}

// NewDeduplicator creates a deduplicator. Pass nil for embedder to
// skip the semantic tier and run exact + fuzzy only.
// Deprecated: Use NewDeduplicatorWithConfig for configurable thresholds.
func NewDeduplicator(embedder llm.Embedder) *Deduplicator {
	return NewDeduplicatorWithConfig(DeduplicatorConfig{Embedder: embedder})
}

// NewDeduplicatorWithConfig creates a Deduplicator with configurable
// thresholds and similarity algorithm. Zero-value fields use defaults.
func NewDeduplicatorWithConfig(cfg DeduplicatorConfig) *Deduplicator {
	fuzzy := cfg.FuzzyThreshold
	if fuzzy == 0 {
		fuzzy = FuzzyLabelThreshold
	}
	autoMerge := cfg.AutoMergeThreshold
	if autoMerge == 0 {
		autoMerge = EmbeddingAutoMerge
	}
	review := cfg.ReviewThreshold
	if review == 0 {
		review = EmbeddingReviewThreshold
	}
	sim := cfg.Similarity
	if sim == nil {
		sim = JaroWinklerDistance
	}
	return &Deduplicator{
		embedder:        cfg.Embedder,
		fuzzyThreshold:  fuzzy,
		autoMerge:       autoMerge,
		reviewThreshold: review,
		similarity:      sim,
	}
}

// Deduplicate compares extracted types against existing types using
// three tiers: exact ID match, fuzzy label match, and semantic
// embedding match.
//
// Time: O(E*X) where E = len(extracted), X = len(existing) for
// exact+fuzzy tiers; semantic tier adds embedding cost.
func (d *Deduplicator) Deduplicate(ctx context.Context, extracted, existing []TypeDefinition) (DeduplicationResult, error) {
	result := DeduplicationResult{
		Unique:           make([]TypeDefinition, 0),
		AutoMerged:       make([]MergedPair, 0),
		ReviewCandidates: make([]MergedPair, 0),
	}

	if len(extracted) == 0 {
		return result, nil
	}
	if len(existing) == 0 {
		result.Unique = append(result.Unique, extracted...)
		return result, nil
	}

	existingByID := buildTypeIDIndex(existing)
	existingByPrefix := groupByPrefix(existing)

	afterExact := d.deduplicateExact(extracted, existingByID, &result)
	afterFuzzy := d.deduplicateFuzzy(afterExact, existingByPrefix, &result)

	if d.embedder == nil || len(afterFuzzy) == 0 {
		result.Unique = append(result.Unique, afterFuzzy...)
		return result, nil
	}

	if err := d.deduplicateSemantic(ctx, afterFuzzy, existing, &result); err != nil {
		return DeduplicationResult{}, fmt.Errorf("ontology: semantic dedup: %w", err)
	}
	return result, nil
}

func (d *Deduplicator) deduplicateExact(
	extracted []TypeDefinition,
	existingByID map[string]TypeDefinition,
	result *DeduplicationResult,
) []TypeDefinition {
	remaining := make([]TypeDefinition, 0, len(extracted))
	for _, entry := range extracted {
		if match, found := existingByID[entry.Type]; found {
			result.AutoMerged = append(result.AutoMerged, MergedPair{
				Extracted:     entry,
				ExistingMatch: match,
				Similarity:    1.0,
				Method:        DedupMethodExact,
			})
			continue
		}
		remaining = append(remaining, entry)
	}
	return remaining
}

func (d *Deduplicator) deduplicateFuzzy(
	candidates []TypeDefinition,
	existingByPrefix map[string][]TypeDefinition,
	result *DeduplicationResult,
) []TypeDefinition {
	remaining := make([]TypeDefinition, 0, len(candidates))
	for _, entry := range candidates {
		prefix := typePrefix(entry.Type)
		samePrefix := existingByPrefix[prefix]
		if len(samePrefix) == 0 {
			remaining = append(remaining, entry)
			continue
		}

		matched := false
		for _, existing := range samePrefix {
			sim := d.similarity(entry.Label, existing.Label)
			if sim >= d.fuzzyThreshold {
				result.AutoMerged = append(result.AutoMerged, MergedPair{
					Extracted:     entry,
					ExistingMatch: existing,
					Similarity:    sim,
					Method:        DedupMethodFuzzyLabel,
				})
				matched = true
				break
			}
		}

		if !matched {
			remaining = append(remaining, entry)
		}
	}
	return remaining
}

func (d *Deduplicator) deduplicateSemantic(
	ctx context.Context,
	candidates []TypeDefinition,
	existing []TypeDefinition,
	result *DeduplicationResult,
) error {
	candidateTexts := make([]string, len(candidates))
	for i, c := range candidates {
		candidateTexts[i] = c.Label + ": " + c.Description
	}

	existingTexts := make([]string, len(existing))
	for i, e := range existing {
		existingTexts[i] = e.Label + ": " + e.Description
	}

	candidateEmbeddings, err := d.embedder.Embed(ctx, candidateTexts)
	if err != nil {
		return fmt.Errorf("embedding candidates: %w", err)
	}
	existingEmbeddings, err := d.embedder.Embed(ctx, existingTexts)
	if err != nil {
		return fmt.Errorf("embedding existing types: %w", err)
	}

	for i, candidate := range candidates {
		if i >= len(candidateEmbeddings) {
			result.Unique = append(result.Unique, candidate)
			continue
		}
		candidateVec := candidateEmbeddings[i]

		var bestSimilarity float64
		var bestMatch TypeDefinition
		hasBest := false

		for j, existingEntry := range existing {
			if j >= len(existingEmbeddings) {
				break
			}
			similarity, err := CosineSimilarity(candidateVec, existingEmbeddings[j])
			if err != nil {
				return fmt.Errorf("ontology: cosine similarity: %w", err)
			}
			if similarity > bestSimilarity {
				bestSimilarity = similarity
				bestMatch = existingEntry
				hasBest = true
			}
		}

		switch {
		case hasBest && bestSimilarity >= d.autoMerge:
			result.AutoMerged = append(result.AutoMerged, MergedPair{
				Extracted:     candidate,
				ExistingMatch: bestMatch,
				Similarity:    bestSimilarity,
				Method:        DedupMethodEmbedding,
			})
		case hasBest && bestSimilarity >= d.reviewThreshold:
			result.ReviewCandidates = append(result.ReviewCandidates, MergedPair{
				Extracted:     candidate,
				ExistingMatch: bestMatch,
				Similarity:    bestSimilarity,
				Method:        DedupMethodEmbedding,
			})
		default:
			result.Unique = append(result.Unique, candidate)
		}
	}
	return nil
}

// DeduplicateType evaluates a single candidate against a list of
// existing types using three tiers: exact ID match, fuzzy label match,
// and semantic embedding. Pass nil for embedder to skip the semantic tier.
func DeduplicateType(ctx context.Context, candidate TypeDefinition, existing []TypeDefinition, embedder llm.Embedder) (DedupResult, error) {
	if len(existing) == 0 {
		return DedupResult{Kind: DedupResultNew}, nil
	}

	for i := range existing {
		if candidate.Type == existing[i].Type {
			return DedupResult{Kind: DedupResultExact, ExistingType: &existing[i], Similarity: 1.0}, nil
		}
	}

	candidatePrefix := typePrefix(candidate.Type)
	for i := range existing {
		if typePrefix(existing[i].Type) != candidatePrefix {
			continue
		}
		similarity := JaroWinklerDistance(candidate.Label, existing[i].Label)
		if similarity >= FuzzyLabelThreshold {
			return DedupResult{Kind: DedupResultFuzzyMatch, ExistingType: &existing[i], Similarity: similarity}, nil
		}
	}

	if embedder == nil {
		return DedupResult{Kind: DedupResultNew}, nil
	}

	candidateText := candidate.Label + ": " + candidate.Description
	existingTexts := make([]string, len(existing))
	for i, e := range existing {
		existingTexts[i] = e.Label + ": " + e.Description
	}

	candidateEmbeddings, err := embedder.Embed(ctx, []string{candidateText})
	if err != nil {
		return DedupResult{}, fmt.Errorf("ontology: embed candidate: %w", err)
	}
	existingEmbeddings, err := embedder.Embed(ctx, existingTexts)
	if err != nil {
		return DedupResult{}, fmt.Errorf("ontology: embed existing: %w", err)
	}
	if len(candidateEmbeddings) == 0 {
		return DedupResult{Kind: DedupResultNew}, nil
	}

	candidateVec := candidateEmbeddings[0]
	var bestSimilarity float64
	var bestIdx int

	for j := range existingEmbeddings {
		sim, err := CosineSimilarity(candidateVec, existingEmbeddings[j])
		if err != nil {
			return DedupResult{}, fmt.Errorf("ontology: cosine similarity: %w", err)
		}
		if sim > bestSimilarity {
			bestSimilarity = sim
			bestIdx = j
		}
	}

	switch {
	case bestSimilarity >= EmbeddingAutoMerge:
		return DedupResult{Kind: DedupResultSemanticMatch, ExistingType: &existing[bestIdx], Similarity: bestSimilarity}, nil
	case bestSimilarity >= EmbeddingReviewThreshold:
		return DedupResult{Kind: DedupResultSemanticReview, ExistingType: &existing[bestIdx], Similarity: bestSimilarity}, nil
	default:
		return DedupResult{Kind: DedupResultNew}, nil
	}
}

func buildTypeIDIndex(types []TypeDefinition) map[string]TypeDefinition {
	index := make(map[string]TypeDefinition, len(types))
	for _, t := range types {
		index[t.Type] = t
	}
	return index
}

func groupByPrefix(types []TypeDefinition) map[string][]TypeDefinition {
	groups := make(map[string][]TypeDefinition)
	for _, t := range types {
		prefix := typePrefix(t.Type)
		groups[prefix] = append(groups[prefix], t)
	}
	return groups
}

func typePrefix(typeID string) string {
	idx := strings.IndexByte(typeID, '.')
	if idx < 0 {
		return typeID
	}
	return typeID[:idx]
}

File: sdks/ts/memory/src/ontology/dedup.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Three-tier ontology type deduplication: exact ID match, fuzzy label
 * match (Jaro-Winkler >= 0.85 within same prefix), and semantic
 * embedding match (cosine >= 0.9 auto-merge, >= 0.75 review).
 *
 * Port of apps/intelligence-service/src/services/document-processing/
 * ontology-type-deduplicator.ts adapted for the standalone memory
 * package.
 */

import type { Embedder } from '../llm/types.js'
import type { TypeDefinition } from './types.js'

import { cosineSimilarity, jaroWinklerDistance } from './similarity.js'

export const FUZZY_LABEL_THRESHOLD = 0.85
export const EMBEDDING_AUTO_MERGE_THRESHOLD = 0.9
export const EMBEDDING_REVIEW_THRESHOLD = 0.75

export type DedupResultKind = 'exact' | 'fuzzy_match' | 'semantic_match' | 'semantic_review' | 'new'

export type DedupResult = {
  readonly kind: DedupResultKind
  readonly existingType?: TypeDefinition
  readonly similarity: number
}

export type DedupMethod = 'exact' | 'fuzzy_label' | 'embedding'

export type MergedPair = {
  readonly extracted: TypeDefinition
  readonly existingMatch: TypeDefinition
  readonly similarity: number
  readonly method: DedupMethod
}

export type DeduplicationResult = {
  readonly unique: TypeDefinition[]
  readonly autoMerged: MergedPair[]
  readonly reviewCandidates: MergedPair[]
}

/**
 * A pluggable string similarity function. Accepts two strings and
 * returns a similarity value in [0, 1].
 */
export type SimilarityFn = (a: string, b: string) => number

export type DeduplicatorConfig = {
  readonly embedder?: Embedder
  /** Minimum string similarity for fuzzy label matching. Default: 0.85. */
  readonly fuzzyThreshold?: number
  /** Minimum cosine similarity for automatic semantic merging. Default: 0.9. */
  readonly autoMergeThreshold?: number
  /** Minimum cosine similarity to flag for human review. Default: 0.75. */
  readonly reviewThreshold?: number
  /** Custom string similarity function. Default: jaroWinklerDistance. */
  readonly similarity?: SimilarityFn
}

/**
 * Deduplicator performs three-tier type deduplication against an
 * existing type registry. Thresholds and similarity algorithm are
 * configurable via DeduplicatorConfig.
 *
 * Time: O(E*X) for exact+fuzzy tiers; semantic tier adds embedding
 * cost proportional to batch sizes.
 */
export class Deduplicator {
  private readonly embedder: Embedder | undefined
  private readonly fuzzyThreshold: number
  private readonly autoMergeThreshold: number
  private readonly reviewThreshold: number
  private readonly similarityFn: SimilarityFn

  constructor(config: DeduplicatorConfig) {
    this.embedder = config.embedder
    this.fuzzyThreshold = config.fuzzyThreshold ?? FUZZY_LABEL_THRESHOLD
    this.autoMergeThreshold = config.autoMergeThreshold ?? EMBEDDING_AUTO_MERGE_THRESHOLD
    this.reviewThreshold = config.reviewThreshold ?? EMBEDDING_REVIEW_THRESHOLD
    this.similarityFn = config.similarity ?? jaroWinklerDistance
  }

  async deduplicate(
    extracted: readonly TypeDefinition[],
    existing: readonly TypeDefinition[],
    signal?: AbortSignal,
  ): Promise<DeduplicationResult> {
    const unique: TypeDefinition[] = []
    const autoMerged: MergedPair[] = []
    const reviewCandidates: MergedPair[] = []

    if (extracted.length === 0) {
      return { unique, autoMerged, reviewCandidates }
    }
    if (existing.length === 0) {
      unique.push(...extracted)
      return { unique, autoMerged, reviewCandidates }
    }

    const existingByID = buildTypeIDIndex(existing)
    const existingByPrefix = groupByPrefix(existing)

    const afterExact = deduplicateExact(extracted, existingByID, autoMerged)
    const afterFuzzy = deduplicateFuzzy(afterExact, existingByPrefix, autoMerged, this.fuzzyThreshold, this.similarityFn)

    if (this.embedder === undefined || afterFuzzy.length === 0) {
      unique.push(...afterFuzzy)
      return { unique, autoMerged, reviewCandidates }
    }

    await this.deduplicateSemantic(afterFuzzy, existing, unique, autoMerged, reviewCandidates, signal)
    return { unique, autoMerged, reviewCandidates }
  }

  private async deduplicateSemantic(
    candidates: readonly TypeDefinition[],
    existing: readonly TypeDefinition[],
    unique: TypeDefinition[],
    autoMerged: MergedPair[],
    reviewCandidates: MergedPair[],
    signal?: AbortSignal,
  ): Promise<void> {
    const candidateTexts = candidates.map((c) => `${c.label}: ${c.description}`)
    const existingTexts = existing.map((e) => `${e.label}: ${e.description}`)

    const [candidateEmbeddings, existingEmbeddings] = await Promise.all([
      this.embedder!.embed(candidateTexts, signal),
      this.embedder!.embed(existingTexts, signal),
    ])

    for (let i = 0; i < candidates.length; i++) {
      const candidate = candidates[i]!
      const candidateVec = candidateEmbeddings[i]
      if (candidateVec === undefined) {
        unique.push(candidate)
        continue
      }

      let bestSimilarity = 0
      let bestMatch: TypeDefinition | undefined

      for (let j = 0; j < existing.length; j++) {
        const existingVec = existingEmbeddings[j]
        if (existingVec === undefined) continue
        const similarity = cosineSimilarity(candidateVec, existingVec)
        if (similarity > bestSimilarity) {
          bestSimilarity = similarity
          bestMatch = existing[j]
        }
      }

      if (bestMatch !== undefined && bestSimilarity >= this.autoMergeThreshold) {
        autoMerged.push({
          extracted: candidate,
          existingMatch: bestMatch,
          similarity: bestSimilarity,
          method: 'embedding',
        })
      } else if (bestMatch !== undefined && bestSimilarity >= this.reviewThreshold) {
        reviewCandidates.push({
          extracted: candidate,
          existingMatch: bestMatch,
          similarity: bestSimilarity,
          method: 'embedding',
        })
      } else {
        unique.push(candidate)
      }
    }
  }
}

function deduplicateExact(
  extracted: readonly TypeDefinition[],
  existingByID: Map<string, TypeDefinition>,
  autoMerged: MergedPair[],
): TypeDefinition[] {
  const remaining: TypeDefinition[] = []
  for (const entry of extracted) {
    const match = existingByID.get(entry.type)
    if (match !== undefined) {
      autoMerged.push({
        extracted: entry,
        existingMatch: match,
        similarity: 1.0,
        method: 'exact',
      })
      continue
    }
    remaining.push(entry)
  }
  return remaining
}

function deduplicateFuzzy(
  candidates: readonly TypeDefinition[],
  existingByPrefix: Map<string, TypeDefinition[]>,
  autoMerged: MergedPair[],
  threshold: number = FUZZY_LABEL_THRESHOLD,
  similarityFn: SimilarityFn = jaroWinklerDistance,
): TypeDefinition[] {
  const remaining: TypeDefinition[] = []
  for (const entry of candidates) {
    const prefix = typePrefix(entry.type)
    const samePrefix = existingByPrefix.get(prefix)
    if (samePrefix === undefined || samePrefix.length === 0) {
      remaining.push(entry)
      continue
    }

    let matched = false
    for (const existingEntry of samePrefix) {
      const similarity = similarityFn(entry.label, existingEntry.label)
      if (similarity >= threshold) {
        autoMerged.push({
          extracted: entry,
          existingMatch: existingEntry,
          similarity,
          method: 'fuzzy_label',
        })
        matched = true
        break
      }
    }

    if (!matched) {
      remaining.push(entry)
    }
  }
  return remaining
}

function buildTypeIDIndex(types: readonly TypeDefinition[]): Map<string, TypeDefinition> {
  const index = new Map<string, TypeDefinition>()
  for (const t of types) {
    index.set(t.type, t)
  }
  return index
}

function groupByPrefix(types: readonly TypeDefinition[]): Map<string, TypeDefinition[]> {
  const groups = new Map<string, TypeDefinition[]>()
  for (const t of types) {
    const prefix = typePrefix(t.type)
    const group = groups.get(prefix)
    if (group !== undefined) {
      group.push(t)
    } else {
      groups.set(prefix, [t])
    }
  }
  return groups
}

function typePrefix(typeID: string): string {
  const dotIndex = typeID.indexOf('.')
  if (dotIndex < 0) return typeID
  return typeID.slice(0, dotIndex)
}

/**
 * Evaluate a single candidate type against a list of existing types
 * using three tiers: exact ID match, fuzzy label match (Jaro-Winkler
 * >= 0.85 within same prefix), and semantic embedding match (cosine
 * >= 0.9 auto-merge, >= 0.75 review). Pass undefined for embedder to
 * skip the semantic tier.
 *
 * Time: O(X) for exact+fuzzy; semantic tier adds embedding cost.
 */
export async function deduplicateType(
  candidate: TypeDefinition,
  existing: readonly TypeDefinition[],
  embedder?: Embedder,
  signal?: AbortSignal,
): Promise<DedupResult> {
  if (existing.length === 0) {
    return { kind: 'new', similarity: 0 }
  }

  for (const entry of existing) {
    if (candidate.type === entry.type) {
      return { kind: 'exact', existingType: entry, similarity: 1.0 }
    }
  }

  const candidatePrefix = typePrefix(candidate.type)
  for (const entry of existing) {
    if (typePrefix(entry.type) !== candidatePrefix) continue
    const similarity = jaroWinklerDistance(candidate.label, entry.label)
    if (similarity >= FUZZY_LABEL_THRESHOLD) {
      return { kind: 'fuzzy_match', existingType: entry, similarity }
    }
  }

  if (embedder === undefined) {
    return { kind: 'new', similarity: 0 }
  }

  const candidateText = `${candidate.label}: ${candidate.description}`
  const existingTexts = existing.map((e) => `${e.label}: ${e.description}`)

  const [candidateEmbeddings, existingEmbeddings] = await Promise.all([
    embedder.embed([candidateText], signal),
    embedder.embed(existingTexts, signal),
  ])

  const candidateVec = candidateEmbeddings[0]
  if (candidateVec === undefined) {
    return { kind: 'new', similarity: 0 }
  }

  let bestSimilarity = 0
  let bestEntry: TypeDefinition | undefined

  for (let j = 0; j < existing.length; j++) {
    const existingVec = existingEmbeddings[j]
    if (existingVec === undefined) continue
    const sim = cosineSimilarity(candidateVec, existingVec)
    if (sim > bestSimilarity) {
      bestSimilarity = sim
      bestEntry = existing[j]
    }
  }

  if (bestEntry !== undefined && bestSimilarity >= EMBEDDING_AUTO_MERGE_THRESHOLD) {
    return { kind: 'semantic_match', existingType: bestEntry, similarity: bestSimilarity }
  }
  if (bestEntry !== undefined && bestSimilarity >= EMBEDDING_REVIEW_THRESHOLD) {
    return { kind: 'semantic_review', existingType: bestEntry, similarity: bestSimilarity }
  }
  return { kind: 'new', similarity: 0 }
}
